Fixed all warnings, added "Show" functionality in Notifications

saagarjha opened the pull request:
Created stub methods to satisfy protocols
Fixed a selector
Make Notifications more useful by making the "Show" button in banner alerts actually show the active application
![screen shot 2016-01-03 at 21 42 14](https://cloud.githubusercontent.com/assets/13786931/12084204/08d33748-b265-11e5-8fd5-e91fdd45397d.png)

Stillness-2 commented on BeardedSpice/Preferences/MediaControllerObject.m:
```diff
@@ -17,7 +17,7 @@ - (id)initWithObject:(id)object{
             _name = [object displayName];
         }
-        if ([[object class] instancesRespondToSelector:@selector(isPlaying)] || [[object class] instancesRespondToSelector:@selector(isPlaying:)]) {
+        if ([[object class] instancesRespondToSelector:@selector(isPlaying)] || [[object class] instancesRespondToSelector:@selector(isPlaying)]) {
```
what

Stillness-2 commented on BeardedSpice/MediaStrategy.m:
```diff
@@ -68,6 +68,11 @@ -(Track *) trackInfo:(TabAdapter *)tab
+-(BOOL) isPlaying:(TabAdapter *)tab {
+    return NO;
+}
+
```
this is wrong

Stillness-2 commented on BeardedSpice/AppDelegate.m:
```diff
@@ -151,6 +151,13 @@ - (BOOL)userNotificationCenter:(NSUserNotificationCenter *)center shouldPresentN
+- (void)userNotificationCenter:(NSUserNotificationCenter *)center
+       didActivateNotification:(NSUserNotification *)notification {
+    // Raise the currently active tab
+    [activeTab activateTab];
```
May be
```
[self autoSelectedTabs];
[activeTab toggleTab];
```

Stillness-2 commented:
We have hotkey, which raises active app and active tab.  So you propose to duplicate the functionality.
Also in your xcode has its own formatting options that are different. Therefore, PR includes changes, where there is no change.

Stillness-2 commented on BeardedSpice/Tabs/NativeAppTabAdapter.m:
```diff
@@ -27,6 +27,30 @@ + (NSString *)bundleId{
+- (BOOL)isPlaying {
+    return NO;
+}
+
```
This is wrong

saagarjha commented:
Ok, let's go through it:
> We have hotkey, which raises active app and active tab.

This is the most important, and the main reason for this PR. I am aware of the hotkey functionality, and I'm not duplicating it. It's just that the Bearded Spice Notification has a button that says "Show", and a user would expect it to "Show" whatever the Notification is about; in this case the currently playing track. I've simply added that functionality; this is *not* a duplicate of current functionality.

> This is wrong

Could you explain this more please? The warnings were bothering me, do you not want the `TabAdapter` protocol's methods implemented? Do you want other default values?

> what

Was resolving warnings, trying to see what Xcode was complaining about. Did not realize I left it like, without fixing it. Is suppressed in 6cbfba0.

> May be
   `[self autoSelectedTabs];`
   `[activeTab toggleTab];`

Seems to have a lot of stuff setting it to the selected tab and whatnot. When the notification is presented and the user responds, the selected tab shouldn't change. Thus, I thought that directly raising the tab should be enough.

> Also in your xcode has its own formatting options that are different.

I have tried to make my formatting match in 03839ed. Please let me know if where I need to change the formatting if necessary.

Stillness-2 commented:
>Could you explain this more please? 

Check Preferences -> Media Controllers table in your case. All rows contain icon 'A', but this wrong. You implemented `isPlaying` in all classes of the media strategies, including old strategies that do not support it.

Stillness-2 commented:
>This is the most important, and the main reason for this PR. I am aware of the hotkey functionality, and I'm not duplicating it. It's just that the Bearded Spice Notification has a button that says "Show", and a user would expect it to "Show" whatever the Notification is about; in this case the currently playing track. I've simply added that functionality; this is not a duplicate of current functionality.

Have an idea to make a action on a notification, that to find the current track in famous search engines :)

Stillness-2 commented:
>Please let me know if where I need to change the formatting if necessary.

If you looking for `AppDelegate.m` in PR, you will see many changes, for example in the method `setupFavoriteShortcutCallback` and so on. But i do not believe that they exists.

Stillness-2 commented:
>Thus, I thought that directly raising the tab should be enough.

No, because you may move tabs into Safari window before this, for example.

saagarjha commented:
Ok, I've reverted and fixed it based on your suggestions.
The formatting should be back to normal.
I've also removed the stub methods and replaced them with warning suppressions. While I still think this is unsafe (wouldn't calling them cause a crash? How would a client know not to call them?) it appears that certain classes rely upon selectors and their existence was messing them up.
Finally, I've removed the raising functionality. I'm not quite sure what you mean by 

> find the current track in famous search engines

Do you want to open a page on Google with the track name? Search for it on iTunes?

Stillness-2 commented:
> Do you want to open a page on Google with the track name? Search for it on iTunes?

Yes, kind of

Stillness-2 commented:
We need tab in Preferences, in which user may select appropriate search engine for him.

saagarjha commented:
Sure, I'll work on it when I have time. Do you have any way you want it implemented?

Stillness-2 commented:
Sorry, but i close this PR. I take #pragma commands from here. :)
